pkg/resolver/related.go

```diff
@@ -22,3 +32,328 @@ type SearchRelatedResult struct {
+	// add a union to the relation query to get all resources in the clusters
+	clusterSelectTerm := s.selectIfClusterUIDPresent()
+	if clusterSelectTerm != nil {
+		relQuery = relQuery.Union(clusterSelectTerm)
+	}
+	sql, params, err := relQuery.ToSQL()
+
+	if err != nil {
+		klog.Error("Error creating relation query", err)
+	} else {
+		s.query = sql
+		s.params = params
+		klog.V(3).Info("Relations query: ", s.query)
+	}
+}
+
+// Check if clusters are part of the search input `kind: Cluster`
+func (s *SearchResult) selectIfClusterUIDPresent() *goqu.SelectDataset {
+	var clusterNames []string
+	for _, uid := range s.uids { // check if cluster uid is in s.uids
+		if strings.HasPrefix(*uid, "cluster__") {
+			clusterName := strings.TrimPrefix(*uid, "cluster__") // get the cluster name
+			clusterNames = append(clusterNames, clusterName)
+		}
+	}
+	if len(clusterNames) > 0 {
+		// Sample query: select uid as uid, data->>'kind' as kind, 1 AS "level" FROM search.resources
+		// where cluster IN ('local-cluster', 'sv-remote-1')
+		//define schema table:
+		schemaTable := goqu.S("search").Table("resources")
+		ds := goqu.From(schemaTable)
+
+		//SELECT CLAUSE
+		selectDs := ds.Select(goqu.C("uid").As("uid"), goqu.L("data->>'kind'").As("kind"), goqu.L("1").As("level"))
+
+		//WHERE CLAUSE - Do we need to add clauses here?
+
+		//LIMIT CLAUSE - Do we need limit here?
+		// limit := config.Cfg.QueryLimit
+
+		return selectDs.Where(goqu.C("cluster").In(clusterNames))
+	} else {
+		return nil
+	}
+}
+
+func (s *SearchResult) buildRelatedKindsQuery() {
+	klog.V(3).Infof("Resolving relationships for [%d] uids.\n", len(s.uids))
+	var params []interface{}
+	var sql string
+	var err error
+	//define schema table:
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+
+	//SELECT CLAUSE
+	selectDs := ds.Select("uid", "cluster", "data")
+
+	//WHERE CLAUSE
+	whereDs := []exp.Expression{goqu.C("uid").In(s.uids)} // Add filter to avoid selecting the search object itself
+
+	//LIMIT CLAUSE
+	limit := s.setLimit()
+
+	//Get the query
+	if limit != 0 {
+		sql, params, err = selectDs.Where(whereDs...).Limit(uint(limit)).ToSQL()
+	} else {
+		sql, params, err = selectDs.Where(whereDs...).ToSQL()
+	}
+	if err != nil {
+		klog.Errorf("Error building SearchRelatedKinds query: %s", err.Error())
+		return
+	}
+	klog.V(3).Infof("SearchRelatedKinds query: %s\nargs: %s", sql, params)
+	s.query = sql
+	s.params = params
+}
+
+func (s *SearchResult) getRelations() []SearchRelatedResult {
+	klog.V(3).Infof("Resolving relationships for [%d] uids.\n", len(s.uids))
+
+	//defining variables
+	relatedMap := map[string][]string{} // Map to store relations
+
+	// Build the relations query
+	s.buildRelationsQuery()
+
+	relations, relQueryError := s.pool.Query(context.TODO(), s.query, s.params...) // how to deal with defaults.
+	if relQueryError != nil {
+		klog.Errorf("Error while executing getRelations query. Error :%s", relQueryError.Error())
+		return nil
```

Comment:
Maybe it's safer to return an empty array instead of nil. We had a similar bug where the nil value broke the UI.

Reply:
Done